javaslang/src/main/java/javaslang/collection/Collections.java

```diff
@@ -37,14 +37,32 @@
         return (C) collection.filter(removed::contains);
     }
 
+    public static <T, C, R extends Iterable<T>> Map<C, R> groupBy(Traversable<T> collection, Function<? super T, ? extends C> classifier, Function<? super Iterable<T>, R> mapper) {
```

Comment:
extracted, in order for the mutable-to-persistent transformation only happens once